java/lance-jni/src/blocking_dataset.rs

```diff
@@ -0,0 +1,36 @@
+use arrow::ffi_stream::ArrowArrayStreamReader;
+use tokio::runtime::Runtime;
+
+use lance::Dataset;
+use lance::Result;
+
+pub struct BlockingDataset {
+    inner: Dataset,
+    rt: Runtime,
+}
+
+impl BlockingDataset {
+    pub fn write(reader: ArrowArrayStreamReader, uri: &str) -> Result<Self> {
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()?;
+
+        let inner = rt.block_on(
+            Dataset::write(reader, uri, Option::None)
+        )?;
+        Ok(BlockingDataset { inner, rt })
+    }
+    pub fn open(uri: &str) -> Result<Self> {
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()?;
+
+        let inner = rt.block_on(Dataset::open(uri))?;
+        Ok(BlockingDataset { inner, rt })
+    }
+
+    pub fn count_rows(&self) -> Result<usize> {
+        self.rt.block_on(self.inner.count_rows())
+    }
+    pub fn close(&self) {}
```

Comment:
So input and output are `FFI_ArrowArrayStream` iirc?

Reply:
I think the input of the write path can be either a FFI_ArrowArrayStream or FFI_ArrowArray (I have implement the stream api for write,  I'm thinking to implement append api for both structure).  For the output of the read path, I think it will be something very similar but I didn't get a chance to try to implement it.